engine/math/matrix.lisp

```diff
@@ -1,7 +1,238 @@
+
+
+(defmethod get-at ((matrix float2x2) i j)
+  (aref (in-list matrix) (+ i (* j 2))))
+
+(defmethod get-at ((matrix float3x3) i j)
+  (aref (in-list matrix) (+ i (* j 3))))
+
+(defmethod get-at ((matrix float4x4) i j)
+  (aref (in-list matrix) (+ i (* j 4))))
+
+
+(defmethod set-at ((matrix float2x2) i j v)
+  (setf (aref (in-list matrix) (+ i (* j 2))) v))
+
+(defmethod set-at ((matrix float3x3) i j v)
+  (setf (aref (in-list matrix) (+ i (* j 3))) v))
+
+(defmethod set-at ((matrix float4x4) i j v)
+  (setf (aref (in-list matrix) (+ i (* j 4))) v))
+
+
+(defmethod add ((matrix1 float2x2) (matrix2 float2x2))
+  (make-instance
+   'float2x2
+   :in-list (map 'vector
+	     (lambda (x y)
+	       (+ x y))
+	     (in-list matrix1)
+	     (in-list matrix2))))
+
+(defmethod add ((matrix1 float3x3) (matrix2 float3x3))
+  (make-instance
+   'float3x3
+   :in-list (map 'vector
+	     (lambda (x y)
+	       (+ x y))
+	     (in-list matrix1)
+	     (in-list matrix2))))
+
+(defmethod add ((matrix1 float4x4) (matrix2 float4x4))
+  (make-instance
+   'float4x4
+   :in-list (map 'vector
+	     (lambda (x y)
+	       (+ x y))
+	     (in-list matrix1)
+	     (in-list matrix2))))
+
+
+(defmethod sub ((matrix1 float2x2) (matrix2 float2x2))
+  (make-instance
+   'float2x2
+   :in-list (map 'vector
+	     (lambda (x y)
+	       (- x y))
+	     (in-list matrix1)
+	     (in-list matrix2))))
+
+(defmethod sub ((matrix1 float3x3) (matrix2 float3x3))
+  (make-instance
+   'float3x3
+   :in-list (map 'vector
+	     (lambda (x y)
+	       (- x y))
+	     (in-list matrix1)
+	     (in-list matrix2))))
+
+(defmethod sub ((matrix1 float4x4) (matrix2 float4x4))
+  (make-instance
+   'float4x4
+   :in-list (map 'vector
+	     (lambda (x y)
+	       (- x y))
+	     (in-list matrix1)
+	     (in-list matrix2))))
+
+
+(defmethod mul ((matrix float2x2) scalar)
+  (make-instance
+   'float2x2
+   :in-list (map 'vector
+	     (lambda (x)
+	       (* x scalar))
+	     (in-list matrix))))
+
+(defmethod mul ((matrix float3x3) scalar)
+  (make-instance
+   'float3x3
+   :in-list (map 'vector
+	     (lambda (x)
+	       (* x scalar))
+	     (in-list matrix))))
+
+(defmethod mul ((matrix float4x4) scalar)
+  (make-instance
+   'float4x4
+   :in-list (map 'vector
+	     (lambda (x)
+	       (* x scalar))
+	     (in-list matrix))))
+
+
+(defmethod div ((matrix float2x2) scalar)
+  (make-instance
+   'float2x2
+   :in-list (map 'vector
+	     (lambda (x)
+	       (/ x scalar))
+	     (in-list matrix))))
+
+(defmethod div ((matrix float3x3) scalar)
+  (make-instance
+   'float3x3
+   :in-list (map 'vector
+	     (lambda (x)
+	       (/ x scalar))
+	     (in-list matrix))))
+
+(defmethod div ((matrix float4x4) scalar)
+  (make-instance
+   'float4x4
+   :in-list (map 'vector
+	     (lambda (x)
+	       (/ x scalar))
+	     (in-list matrix))))
+
+
+(defmethod prod ((matrix1 float2x2) (matrix2 float2x2))
+  (let ((outm (make-instance 'float2x2)))
+    (loop
```

Comment:
That indentation should be fixed into something more readable.
Like this or so: 
```cl
(loop for i from 0 to 1 do
  (loop for j from 0 to 1 do
    (set-at outm i j
            (reduce #'+
                    (mapcar #'* (get-col matrix i)
                            (get-row value j))))))
```

Reply:
I think I'll just use `dotimes` instead of loop in this case since it's more clear.